Support complex number list in JIT

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51145 Support complex number list in JIT**

Differential Revision: [D26154025](https://our.internmc.facebook.com/intern/diff/D26154025)

anjali411 committed:
Support complex number list in JIT

[ghstack-poisoned]

SplitInfinity commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1120,12 +1124,12 @@ std::string ClassType::getForwardPreHookErrorMessage(int pre_hook_idx) const {
   const FunctionSchema& forward_schema = getMethod("forward").getSchema();
   std::string input_types = getSchemaInputTypesString(forward_schema);
   const std::vector<Argument>& forward_args = forward_schema.arguments();
-   
```
There seem to be a lot of formatting changes below this point. Can you undo them?

anjali411 replied:
they seem inevitable

anjali411 replied:
done!

SplitInfinity reviewed (commented):
I'm not sure if this is ready for review (I didn't check if you requested review already, oops) but there are a few distracting formatting changes, and there should be Python tests that use lists of complex numbers.

anjali411 committed:
Update on "Support complex number list in JIT"




[ghstack-poisoned]

anjali411 committed:
Update on "Support complex number list in JIT"




[ghstack-poisoned]

anjali411 committed:
Update on "Support complex number list in JIT"




[ghstack-poisoned]

anjali411 commented:
@SplitInfinity ready for review. Added the python test, but can't find a way to update that file without causing those formatting changes

anjali411 committed:
Update on "Support complex number list in JIT"




[ghstack-poisoned]

anjali411 committed:
Update on "Support complex number list in JIT"




[ghstack-poisoned]

SplitInfinity commented on aten/src/ATen/test/ivalue_test.cpp:
```diff
@@ -49,6 +49,26 @@ TEST(IValueTest, Basic) {
+  ASSERT_EQ(foo12.toComplexDouble(), c10::complex<double>(3,4));
+  ASSERT_EQ(foo1.use_count(), 2);
+  ASSERT_TRUE(baz1.toComplexDoubleVector() == std::vector<c10::complex<double>>({(3, 4), (3, -4), (5, 0)}));
+  IValue the_complex_list(
```
Seems a bit weird to call this a list when it is a tuple.

anjali411 replied:
true updated the name!

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -14,6 +14,13 @@ def fn(a: complex):
 
         self.checkScript(fn, (3 + 5j,))
 
+    def test_complexlist(self):
+        def fn(a: List[complex], idx: int):
+            return a[idx]
+
+        input = [1j, 2, 3 + 4j, -5, -7j]
```
What about serialization of lists? Is that coming later?

anjali411 replied:
will add it in a follow-up PR

SplitInfinity reviewed (commented):
Nice!

SplitInfinity commented on torch/csrc/jit/ir/constants.cpp:
```diff
@@ -88,8 +88,8 @@ c10::optional<Value*> tryInsertConstant(
   } else if (val.isList()) {
-    bool fast_path_list =
-        val.isBoolList() || val.isIntList() || val.isDoubleList();
+    bool fast_path_list = val.isBoolList() || val.isIntList() ||
+        val.isDoubleList() || val.isComplexDoubleList();
```
As I recall, complex literals are not supported in TorchScript yet, so this is not being tested.

anjali411 replied:
removed!

anjali411 committed:
Update on "Support complex number list in JIT"


Differential Revision: [D26154025](https://our.internmc.facebook.com/intern/diff/D26154025)

[ghstack-poisoned]

SplitInfinity reviewed (approved):
Nice

anjali411 committed:
Update on "Support complex number list in JIT"


Differential Revision: [D26154025](https://our.internmc.facebook.com/intern/diff/D26154025)

[ghstack-poisoned]

anjali411 committed:
Update on "Support complex number list in JIT"


Differential Revision: [D26154025](https://our.internmc.facebook.com/intern/diff/D26154025)

[ghstack-poisoned]
